File: ios/Classes/EventPlugin.m
//
//  EventPlugin.m
//  Airship
//
//  Created by Ulrich Giberne on 17/05/2021.
//

//#import "EventPlugin.h"
//
//#import <Foundation/Foundation.h>
//
//@implementation EventPlugin {
//    FlutterEngine *_headlessRunner;
//    FlutterMethodChannel *_callbackChannel;
//    FlutterMethodChannel *_mainChannel;
//    NSObject<FlutterPluginRegistrar> *_registrar;
//    NSUserDefaults *_persistentState;
//    NSMutableArray *_eventQueue;
//}
//
//static EventPlugin *instance = nil;
//static FlutterPluginRegistrantCallback registerPlugins = nil;
////static BOOL initialized = NO;
//static BOOL backgroundIsolateRun = NO;
//
//+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
//  @synchronized(self) {
//    if (instance == nil) {
//      NSLog(@"Registering with registrar");
//      instance = [[EventPlugin alloc] init:registrar];
//      [registrar addApplicationDelegate:instance];
//    }
//  }
//}
//
//- (instancetype)init:(NSObject<FlutterPluginRegistrar> *)registrar {
//    self = [super init];
//    NSAssert(self, @"super init cannot be nil");
//
//    // 1. Retrieve NSUserDefaults which will be used to store callback handles
//    // between launches.
//    _persistentState = [NSUserDefaults standardUserDefaults];
//
//    // 3. Initialize the Dart runner which will be used to run the callback
//    // dispatcher.
//    _headlessRunner = [[FlutterEngine alloc]
//                        initWithName:@"EventIsolate"
//                        project:nil
//                        allowHeadlessExecution:YES];
//    _registrar = registrar;
//
//    // 4. Create the method channel used by the Dart interface to invoke
//    // methods and register to listen for method calls.
//    _mainChannel = [FlutterMethodChannel
//                    methodChannelWithName:@"com.airship.flutter/event_plugin"
//                    binaryMessenger:[registrar messenger]];
//    [registrar addMethodCallDelegate:self channel:_mainChannel];
//
//    // 5. Create a second method channel to be used to communicate with the
//    // callback dispatcher. This channel will be registered to listen for
//    // method calls once the callback dispatcher is started.
//    _callbackChannel = [FlutterMethodChannel
//                        methodChannelWithName:@"com.airship.flutter/event_plugin_background"
//                        binaryMessenger:_headlessRunner];
//
//    return self;
//}
//
//- (void)startEventService:(int64_t)handle {
//    NSLog(@"Initializing EventService");
//    [self setCallbackDispatcherHandle:handle];
//    FlutterCallbackInformation *info = [FlutterCallbackCache lookupCallbackInformation:handle];
//    NSAssert(info != nil, @"failed to find callback");
//    NSString *entrypoint = info.callbackName;
//    NSString *uri = info.callbackLibraryPath;
//    [_headlessRunner runWithEntrypoint:entrypoint libraryURI:uri];
//    NSAssert(registerPlugins != nil, @"failed to set registerPlugins");
//
//    // Once our headless runner has been started, we need to register the application's plugins
//    // with the runner in order for them to work on the background isolate. `registerPlugins` is
//    // a callback set from AppDelegate.m in the main application. This callback should register
//    // all relevant plugins (excluding those which require UI).
//    if (!backgroundIsolateRun) {
//        registerPlugins(_headlessRunner);
//    }
//    [_registrar addMethodCallDelegate:self channel:_callbackChannel];
//    backgroundIsolateRun = YES;
//}
//
//- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
//    NSArray *arguments = call.arguments;
//    if ([@"EventPlugin.performAction" isEqualToString:call.method]) {
//        [self startEventService:[arguments[0] longValue]];
//        result(@(YES));
//    } else if ([@"EventService.performed" isEqualToString:call.method]) {
////        @synchronized(self) {
////              initialized = YES;
////                // Send the geofence events that occurred while the background
////                // isolate was initializing.
////                while ([_eventQueue count] > 0) {
////                    NSDictionary* event = _eventQueue[0];
////                    [_eventQueue removeObjectAtIndex:0];
////                    CLRegion* region = [event objectForKey:kRegionKey];
////                    int type = [[event objectForKey:kEventType] intValue];
////                    [self sendLocationEvent:region eventType: type];
////                }
////            }
//        result(nil);
//    } else {
//        result(FlutterMethodNotImplemented);
//    }
//}
//
//- (BOOL)application:(UIApplication *)application
//    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
//
//    // Check to see if we're being launched due to a location event.
//    if (launchOptions[UIApplicationLaunchOptionsLocationKey] != nil) {
//        // Restart the headless service.
//        [self startEventService:[self getCallbackDispatcherHandle]];
//    }
//
//    // Note: if we return NO, this vetos the launch of the application.
//    return YES;
//}
//
//- (int64_t)getCallbackDispatcherHandle {
//    id handle = [_persistentState objectForKey:@"callback_dispatcher_handle"];
//    if (handle == nil) {
//        return 0;
//    }
//    return [handle longLongValue];
//}
//
//- (void)setCallbackDispatcherHandle:(int64_t)handle {
//    [_persistentState setObject:[NSNumber numberWithLongLong:handle] forKey:@"callback_dispatcher_handle"];
//}
//
//@end

File: ios/Classes/EventPlugin.swift
import Flutter
import Foundation

var instance: EventPlugin? = nil
var registerPlugins: FlutterPluginRegistrantCallback? = nil
//static BOOL initialized = NO;
var backgroundIsolateRun = false

//public class EventPlugin : NSObject, FlutterPlugin {
//    private var headlessRunner: FlutterEngine?
//    private var callbackChannel: FlutterMethodChannel?
//    private var mainChannel: FlutterMethodChannel?
//    private weak var registrar: (NSObjectProtocol & FlutterPluginRegistrar)?
//    private var persistentState: UserDefaults?
//    private var eventQueue: [AnyHashable]?
//
//    public class func register(with registrar: (NSObjectProtocol & FlutterPluginRegistrar)) {
//        let lockQueue = DispatchQueue(label: "self")
//        lockQueue.sync {
//            if instance == nil {
//                print("Registering with registrar")
//                instance = EventPlugin(registrar)
//                registrar.addApplicationDelegate(instance!)
//            }
//        }
//    }
//
//    public init(_ test: String) {
//        super.init()
//
//        // 1. Retrieve NSUserDefaults which will be used to store callback handles
//        // between launches.
//        persistentState = UserDefaults.standard
//    }
//
//    init(_ registrar: (NSObjectProtocol & FlutterPluginRegistrar)?) {
//        super.init()
//
//        // 1. Retrieve NSUserDefaults which will be used to store callback handles
//        // between launches.
//        persistentState = UserDefaults.standard
//
//        // 3. Initialize the Dart runner which will be used to run the callback
//        // dispatcher.
//        headlessRunner = FlutterEngine(
//            name: "EventIsolate",
//            project: nil,
//            allowHeadlessExecution: true)
//        self.registrar = registrar
//
//        // 4. Create the method channel used by the Dart interface to invoke
//        // methods and register to listen for method calls.
//        mainChannel = FlutterMethodChannel(
//            name: "com.airship.flutter/event_plugin",
//            binaryMessenger: registrar?.messenger() as! FlutterBinaryMessenger)
//        registrar?.addMethodCallDelegate(self, channel: mainChannel!)
//
//        // 5. Create a second method channel to be used to communicate with the
//        // callback dispatcher. This channel will be registered to listen for
//        // method calls once the callback dispatcher is started.
//        callbackChannel = FlutterMethodChannel(
//                            name:"com.airship.flutter/event_plugin_background",
//            binaryMessenger: headlessRunner as! FlutterBinaryMessenger)
//    }
//
//    // TODO Move things directly into the AirshipPlugin
//    public func startEventService(_ handle: Int64) {
//        print("Initializing EventService")
//        setCallbackDispatcherHandle(handle)
//        let info = FlutterCallbackCache.lookupCallbackInformation(handle)
//        assert(info != nil, "failed to find callback")
//        let entrypoint = info?.callbackName
//        let uri = info?.callbackLibraryPath
//        headlessRunner?.run(withEntrypoint: entrypoint, libraryURI: uri)
//        assert(registerPlugins != nil, "failed to set registerPlugins")
//
//        // Once our headless runner has been started, we need to register the application's plugins
//        // with the runner in order for them to work on the background isolate. `registerPlugins` is
//        // a callback set from AppDelegate.m in the main application. This callback should register
//        // all relevant plugins (excluding those which require UI).
//        if !backgroundIsolateRun {
//            registerPlugins?(headlessRunner!)
//        }
//        registrar?.addMethodCallDelegate(self, channel: callbackChannel!)
//        backgroundIsolateRun = true
//    }
//
//    public func handle(_ call: FlutterMethodCall, result: FlutterResult) {
//        let arguments = call.arguments as? NSArray
//        if "EventPlugin.performAction" == call.method {
//            startEventService(Int64((arguments?[0] as! NSNumber)))
//            result(NSNumber(value: true))
//        } else if "EventService.performed" == call.method {
//
//            result(nil)
//        } else {
//            result(FlutterMethodNotImplemented)
//        }
//    }
//
//    private func application(
//        _ application: UIApplication,
//        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil
//    ) -> Bool {
//
//        // Check to see if we're being launched due to a location event.
//        if launchOptions?[UIApplication.LaunchOptionsKey.location] != nil {
//            // Restart the headless service.
//            startEventService(getCallbackDispatcherHandle())
//        }
//
//        // Note: if we return NO, this vetos the launch of the application.
//        return true
//    }
//
//    func getCallbackDispatcherHandle() -> Int64 {
//        let handle = persistentState?.object(forKey: "callback_dispatcher_handle")
//        if handle == nil {
//            return 0
//        }
//        return (handle as? NSNumber)?.int64Value ?? 0
//    }
//
//    func setCallbackDispatcherHandle(_ handle: Int64) {
//        persistentState?.set(NSNumber(value: handle), forKey: "callback_dispatcher_handle")
//    }
//}
